content/ADR-203-social-ws-login.md

```diff
@@ -0,0 +1,156 @@
+- Implementation is simpler, as there is no need for custom logic on the client side or token management, nor does it have to handle reconnections.
+- Stateless.
+- Compatible if we want to migrate it to HTTP.
+- Supports automatic reconnection or node change (through Load Balancer for example) without forcing a reconnection.
+
+#### Solution 3: Disadvantages
+
+- The implementation is a bit more complex as an additional authentication system and the generation of a JWT by the server is required.
+- If the user changes their token, the `FriendshipToken` must be regenerated. This can generate a bit more overhead on the server.
+- If the user has multiple tokens (for example, on different devices), then they would have to manually close all connections.
+- If the server crashes, all generated tokens are lost, and clients would have to generate new tokens again.
+
+### Solution 4: Authenticate the Websocket Connection
+
+To prioritize minimizing requests to Redis and Synapse while taking advantage of an open connection to send the token only once and not having to send it with each message, another possible solution is to authenticate the websocket connection when it is opened. This way, only one websocket connection with the server is allowed when authenticated, and if no token is sent or it is incorrect, the connection is closed.
+
+To achieve this, the client that opens the connection must send the authentication token via a header:
+
+- The client must make an HTTP request to the server with the authentication header to request an upgrade to a websocket connection.
+- The server validates the token, and if it is correct, allows the connection to be upgraded to a websocket connection and keeps the connection open. If the token is incorrect, the connection is closed.
+- With the connection open, the client can send messages as desired without having to send the token again.
+
+The native browser client does not allow headers to be added when creating a websocket connection. Therefore, the proposed solution is to use cookies:
+
+- The client obtains the token through the social service HTTP endpoint `/login`, where the response includes the header `Set-Cookie: friendships-Authorization=<Token>; Domain=social.decentraland.org`.
+- Then, when the client initializes the websocket, since the cookie is configured, the header `Cookie: friendships-Authorization=<Token>` is sent, and the server validates the token and obtains the user_id for the entire connection.
+Example of proto file:
+
+```proto
+service FriendshipsService {
+  rpc GetFriends(google.protobuf.Empty) returns (stream Users) {}
+}
+```
+
+#### Solution 4: Advantages
+
+- Sends the request to Redis/Synapse only once
+- Does not allow more than one connection per user (prevents simple DOS)
+- No need to send the token as part of the message
+
+#### Solution 4: Disadvantages
+
+- The client must support cookies or be able to send the authorization header.
```

Comment:
There is currently so much struggle while mixing cookies and websockets. It is not a straight forward implementation in _most_ languages.

Besides the ergonomics of getting cookies to work, I think setting cookies is something very web 1.0 for native public-key applications like Decentraland. I'd like to propose the ws-comms authentication mechanism for the WS connection. It is implemented in `bff`, `ws-room-service` and `mini-comms` repositories. 

```mermaid
sequenceDiagram
  client-->>server: open ws
  server-->>client: pleaseSignChallenge(challengeText=randomString())
  client-->>server: signedChallenge(result=sign(challengeText))
  server-->>server: authenticate, authorize
  server-->>server: load modules based on the permissions
  server-->>client: welcomeMessages(avalableModulesList)
```

The messages are simple and straight forward to implement, and the solution without cookies adds an extra layer of security against session spoofing

[BFF proto definitions](https://github.com/decentraland/protocol/blob/main/proto/decentraland/bff/authentication_service.proto#L24-L27)

```protobuf
syntax = "proto3";
package decentraland.bff;

message GetChallengeRequest {
  string address = 1;
}

message GetChallengeResponse {
  string challenge_to_sign = 1;
  bool already_connected = 2;
}

message SignedChallenge {
  string auth_chain_json = 1;
}

message WelcomePeerInformation {
  string peer_id = 1;

  // list of available modules in this BFF
  repeated string available_modules = 2;
}

service BffAuthenticationService {
  rpc GetChallenge(GetChallengeRequest) returns (GetChallengeResponse) {}
  rpc Authenticate(SignedChallenge) returns (WelcomePeerInformation) {}
}
```

The tests are also super straight forward https://github.com/decentraland/explorer-bff/blob/4352737ca24a0c590387fba4b46da297408e899a/test/integration/rpc.spec.ts#L18-L31

This has a clear advantage of not needing to store a second token, only making use of the already available auth chain to authenticate the users. It will also enable interoperability with the [scoped permissions for auth chains](https://github.com/decentraland/adr/pull/162/files)